lom/src/lib/lomcommon/helper.go

```diff
@@ -151,6 +152,100 @@ func LogDebug(s string, a ...interface{})  {
     LogMessage(syslog.LOG_DEBUG, s, a...)
 }
 
+/*
+ * System Shutdown
+ *
+ */
+type sysShutdown_t struct {
+    ch          chan int
+    wg          *sync.WaitGroup
+    shutdown    bool
+}
+
+func (p *sysShutdown_t) register(caller string) <-chan int {
+    if !p.shutdown {
+        LogDebug("sysShutdown_t:Register by (%s)", caller)
+        p.wg.Add(1)
+        return p.ch
+    } else {
+        LogError("Failed to register (%s) as shutdown in progress", caller)
+        return nil
+    }
+}
+
+func (p *sysShutdown_t) deregister(caller string) {
+    LogDebug("sysShutdown_t:Deregister by (%s)", caller)
+    p.wg.Done()
+}
+
+func (p *sysShutdown_t) doShutdown(toutSecs int) {
+    if p.shutdown {
+        LogError("Shutdown already in progress")
+        return
+    }
+    p.shutdown = true
+
+    /* Send alert as many possible */
+    for len(p.ch) < cap(p.ch) {
```

Comment:
Isn't one close(ch) sufficient ? It will signal all waiters on ch . shutdown=1 will prevent any registrations while doShutdown() started.

Reply:
Good one. I will do.

Reply:
Done